### ISSUE TITLE
M68k operand incorrectly decoded causes segfault

### ISSUE THREAD

**Issue by GeoSpark** [aquynh/capstone#1025]
Using 4.0.1024

The following Python script causes a segfault in capstone.

```
from capstone import *
from capstone.m68k import *

print(cs_version())

# move.l  $7e344a.l, $8(a6)
CODE = b'\x2d\x79\x00\x7e\x34\x4a\x00\x08'

# 00
# 10 ; Size long
# 110 101 ; Destination Reg: A6 Mode: (d16,An)
# 111 001 ; Source Mode: (xxx).L  == 11 1001 == 0x39

# 00000000 01111110 00110100 01001010
# 00000000 00001000

# First operand should hit case 0x39 at line 478 of M68KDisassembler.c

md = Cs(CS_ARCH_M68K, CS_MODE_BIG_ENDIAN | CS_MODE_M68K_040)
md.detail = True

for i in md.disasm(CODE, 0x1000):
    print("%x:\t%s\t%s" % (i.address, i.mnemonic, i.op_str))
    if len(i.operands) > 0:
        print("Num operands: %u" % len(i.operands))
        c = -1
        for o in i.operands:
            c += 1

            if o.type == M68K_OP_MEM:
                print("\t\toperand[%u].type: MEM" % c)

                if o.value.mem.base_reg != M68K_REG_INVALID:
                    print("base_reg %x" % o.value.mem.base_reg)
                    print("index_reg %x" % o.value.mem.index_reg)
                    print("in_base_reg %x" % o.value.mem.in_base_reg)
                    print("in_disp %x" % o.value.mem.in_disp)
                    print("out_disp %x" % o.value.mem.out_disp)
                    print("disp %x" % o.value.mem.disp)
                    print("scale %x" % o.value.mem.scale)
                    print("bitfield %x" % o.value.mem.bitfield)
                    print("width %x" % o.value.mem.width)
                    print("offset %x" % o.value.mem.offset)
                    print("index_size %x" % o.value.mem.index_size)
                    print("\t\t\toperand[%u].mem.base: REG = %s" % (c, i.reg_name(o.value.mem.base_reg)))
```

Outputs:

```
(4, 0, 1024)
1000:   move.l  $7e344a.l, $8(a6)
Num operands: 2
                operand[0].type: MEM
base_reg 7e344a
index_reg 0
in_base_reg 0
in_disp 0
out_disp 0
disp 0
scale 0
bitfield 0
width 0
offset 0
index_size 0
Segmentation fault
```

It appears to be interpreting the first operand as a register rather than an absolute address, and so ```reg_name()``` baulks when it tries to access the register names array with a very large index.

Ultimately this breaks radare2's m68k analysis. See: https://github.com/radare/radare2/issues/8663

**Comment by etaba001** [comment]
![]([]())

**Comment by aquynh** [comment]
@emoon & @nplanel, do you have any comments?

**Comment by emoon** [comment]
I will take a look at this later today when I'm home from work.

**Comment by radare** [comment]
tired of waiting. fixed the issue in r2. feel free to apply this patch in capstone

https://github.com/radare/radare2/commit/42ce366601cc4664d74faabbf6768f1ee5303078

**Comment by aquynh** [comment]
can you do pull req?

**Comment by radare** [comment]
yes https://github.com/aquynh/capstone/pull/1027

**Comment by aquynh** [comment]
great, thanks!

@emoon, can you ack?

**Comment by GeoSpark** [comment]
That's great, thanks @radare 

The only thing is that this hides the underlying problem. Not that sanity checks aren't needed, but elsewhere the code is putting the value in the wrong field in the  m68k_op_mem struct (https://github.com/aquynh/capstone/blob/next/include/capstone/m68k.h) The *_reg fields are enums, and I presume the displacement needs to go in the out_disp field.

**Comment by radare** [comment]
I did a quick fix because i need to release without crashes, there must be better ways to fix that probably, that’s up to you and the author of this module.

thanks

**Comment by GeoSpark** [comment]
That's fair enough - I didn't mean to undermine your rapid fix!

**Comment by emoon** [comment]
I would prefer to not have this PR merged as while it hides the crash it doesn't fix the issue as @GeoSpark points out. I will look at this in a few hours.

**Comment by aquynh** [comment]
Agreed that we need a better fix.

**Comment by emoon** [comment]
Proper fix in #1028

**Comment by emoon** [comment]
As a fix was done in #1028 this can be closed I think?

**Comment by aquynh** [comment]
close, thanks
